### PR TITLE
[5.2] Delete user access level - check for levels in use

### PR DESCRIPTION
**Pull request by chmst**
Pull Request replacing https://github.com/joomla/joomla-cms/pull/39654

### Summary of Changes
An access level can only be deleted if there is no content which uses this level. But there is no information given, which tabels are concerned.
This PR adds the list of tables to the error message. It adds the delete method to the model and makes a check for all leves on all tables. 

### Testing Instructions
Add one or more access levels.   
Set this access level for some items in your content, an article, a contact, a module .. whatever.
Then try to delete this access level.

### Actual result BEFORE applying this Pull Request
You get a message
"You can't delete the view access level '%d:%s' because it is being used by content."

### Expected result AFTER applying this Pull Request
grafik
Information is diplayed for all levels, where they are used. 
This enables the experienced user to find the components and filter there for these levels. 

Only leves are deleted (in my test above it was one (1) which are not in use.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by exlemor** [comment]
I have tested this item :white_check_mark: successfully on e449b223be9ba5b5d6e385d1c481d7e12c066172<br /><br />I was able to test this successfully.

Just one tiny correction, the message after Patch is correct unlike the photo: 

(photo) View Access Levels removed ---> (site) No View Access Levels removed.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by chmst** [comment]
In this case I wanted to delete 3 levels. One of them was deleted. Two of them could not be deleted, they are still in use. This is a bit confusing on the screen

**Comment by exlemor** [comment]
Ah ok. In my case, I deleted just 1 level.

File: administrator/components/com_users/src/Model/LevelModel.php [hunk 1]
```diff
@@ -32,10 +32,89 @@
 class LevelModel extends AdminModel
 {
     /**
-     * @var array   A list of the access levels in use.
-     * @since   1.6
+     * Method to delete one or more access levels.
+     *
+     * @param   array  $pks  An array of record primary keys.
```
Comment (brianteeman) [inline administrator/components/com_users/src/Model/LevelModel.php]:
```suggestion
     * @param   array  $pks  An array of primary keys.
```

File: administrator/components/com_users/src/Model/LevelModel.php [hunk 2]
```diff
@@ -32,10 +32,89 @@
...
+        $pks   = (array) $pks;
+        $table = $this->getTable();
+
+        $itemsInUse = [];
+
+        // Iterate the items to delete each one.
+        foreach ($pks as $i => $pk) {
+            if ($table->load($pk)) {
+
+                // Check if the access level is being used by any content.
```
Comment (brianteeman) [inline administrator/components/com_users/src/Model/LevelModel.php]: content ??

File: administrator/components/com_users/src/Model/LevelModel.php [hunk 3]
```diff
@@ -32,10 +32,89 @@
...
+                $db    = $this->getDatabase();
+                $query = $db->getQuery(true)
+                    ->select('DISTINCT access');
+
+                // Get all tabels that have the access field
```
Comment (brianteeman) [inline administrator/components/com_users/src/Model/LevelModel.php]:
```suggestion
                // Get all tables that have the access field
```

File: administrator/components/com_users/src/Model/LevelModel.php [hunk 4]
```diff
@@ -32,10 +32,86 @@
...
+     *
+     * @since   __DEPLOY_VERSION__
      */
-    protected $levelsInUse = null;
+    public function delete(&$pks)
+    {
+        $pks   = (array) $pks;
+        $table = $this->getTable();
+
+        $itemsInUse = [];
```
Comment (Quy) [inline administrator/components/com_users/src/Model/LevelModel.php]:
```suggestion
        $levelsInUse = [];
```

**Comment by cybersalt** [comment]
I have tested this item :white_check_mark: successfully on 615eb4e8d8c67d3b21c0260bf52ab5610abe49b0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by exlemor** [comment]
I have tested this item :white_check_mark: successfully on 615eb4e8d8c67d3b21c0260bf52ab5610abe49b0<br /><br />I have successfully tested it with 2 levels and multiple assets connected to said user levels.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by johnaspr** [comment]
I have tested this item :white_check_mark: successfully on 615eb4e8d8c67d3b21c0260bf52ab5610abe49b0<br /><br />I was able to test this.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by alikon** [comment]
rtc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by joomdonation** [comment]
Sorry, I see few problems with this PR :

- It assumes that we always use `access` field to store access level for the item. In real life (for example, in one of the table in my extension, I uses `registration_access` field to store the access level which can register for the event, and the check is ignored for this field (unfortunately, I don't see any reliable way to handle this)
- What if `access` field in certain table is used for different purpose, I mean not for storing access level? By quick code reading, look like this will prevent the access level from being deleted ?
- The code in the model is building the HTML message and return feedback to users via  `Factory::getApplication()->enqueueMessage` and I think this is not a right way. Ideally, the model method should only do the task and return the result to the controller and then it's up to the controller to decide how to return / display that result to users.

**Comment by chmst** [comment]
@Quy I don't see the remarks as "update requested". 
@Fedik I agree with you. But - I think that your remarks are not in scope of this PR. You could se this as a work-around, because there were conceptual errors in the whole ACL and I cannot heal that in a PR.

**Comment by chmst** [comment]
Changed the code for displaying the message as requested.

**Comment by pe7er** [comment]
Thanks for this PR @chmst 
This PR has been changed, @cybersalt @exlemor @johnaspr Could you please test again?

**Comment by dorisdreher** [comment]
I have tested this item :white_check_mark: successfully on f8ee84c6d52d541eeae36667db760a8fa86e02f9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by KingLouis1** [comment]
I have tested this item :white_check_mark: successfully on f8ee84c6d52d541eeae36667db760a8fa86e02f9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by crimle** [comment]
I have tested this item :white_check_mark: successfully on f8ee84c6d52d541eeae36667db760a8fa86e02f9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

**Comment by MacJoom** [comment]
I have tested this item :white_check_mark: successfully on f8ee84c6d52d541eeae36667db760a8fa86e02f9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43223">issues.joomla.org/tracker/joomla-cms/43223</a>.</sub>

File: administrator/language/en-GB/com_users.ini [hunk 5]
```diff
@@ -104,7 +104,8 @@ COM_USERS_ERROR_LEVELS_NOLEVELS_SELECTED="No View Permission Level(s) selected."
 COM_USERS_ERROR_NO_ADDITIONS="The selected user(s) are already assigned to the selected group."
 COM_USERS_ERROR_NOT_IN_GROUP="The selected user(s) are not in the selected group."
 COM_USERS_ERROR_ONLY_ONE_GROUP="A user must belong to at least one group."
-COM_USERS_ERROR_VIEW_LEVEL_IN_USE="You can't delete the view access level '%d:%s' because it is being used by content."
+COM_USERS_ERROR_VIEW_LEVEL_IN_USE="You can't delete the view access level(s)"
+COM_USERS_ERROR_VIEW_LEVEL_IN_USE_DETAILS="Level with ID %d is being used in the database tables: %s."
```
Comment (brianteeman) [inline administrator/language/en-GB/com_users.ini:108]: can it be more than one table? If not then the string should be in the singular
```suggestion
COM_USERS_ERROR_VIEW_LEVEL_IN_USE_DETAILS="Level with ID %d is being used in the database table: %s."
```

Comment (chmst) [reply administrator/language/en-GB/com_users.ini:108]: it can be more than one, in most cases there are at least two, with finder.

**Pull request merged**